src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorCustomMessageTarget.cs

```diff
@@ -143,4 +163,127 @@ internal partial class RazorCustomMessageTarget
+            if (await _csharpVirtualDocumentAddListener.WaitForDocumentAddAsync(TimeSpan.FromMilliseconds(500), cancellationToken).ConfigureAwait(false))
+            {
+                _outputWindowLogger?.LogDebug("Wait successful!");
+                virtualDocument = FindVirtualDocument<TVirtualDocumentSnapshot>(hostDocument.Uri, hostDocument.GetProjectContext());
+            }
+            else
+            {
+                _outputWindowLogger?.LogDebug("Timed out :(");
+            }
+        }
+
+        if (virtualDocument is null)
+        {
+            _outputWindowLogger?.LogDebug("No virtual document found, falling back to old code.");
+            return await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(requiredHostDocumentVersion, hostDocument.Uri, cancellationToken).ConfigureAwait(false);
+        }
+
+        var result = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(requiredHostDocumentVersion, hostDocument.Uri, virtualDocument.Uri, rejectOnNewerParallelRequest, cancellationToken).ConfigureAwait(false);
+
+        // If we failed to sync on version 1, then it could be that we got new information while waiting, so try again
+        if (requiredHostDocumentVersion == 1 && !result.Synchronized)
+        {
+            _outputWindowLogger?.LogDebug("Sync failed for v1 document. Trying again");
+            virtualDocument = FindVirtualDocument<TVirtualDocumentSnapshot>(hostDocument.Uri, hostDocument.GetProjectContext());
+
+            if (virtualDocument is null)
+            {
+                _outputWindowLogger?.LogDebug("No virtual document found, falling back to old code.");
+                return await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(requiredHostDocumentVersion, hostDocument.Uri, cancellationToken).ConfigureAwait(false);
+            }
+
+            _outputWindowLogger?.LogDebug("Got virtual document after trying again {uri}.Trying again.", virtualDocument.Uri);
+
+            // try again
+            result = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync<TVirtualDocumentSnapshot>(requiredHostDocumentVersion, hostDocument.Uri, virtualDocument.Uri, rejectOnNewerParallelRequest, cancellationToken).ConfigureAwait(false);
+        }
+
+        return result;
+    }
+
+    private SynchronizedResult<TVirtualDocumentSnapshot>? TryReturnPossiblyFutureSnapshot<TVirtualDocumentSnapshot>(
+        int requiredHostDocumentVersion,
+        TextDocumentIdentifier hostDocument) where TVirtualDocumentSnapshot : VirtualDocumentSnapshot
+    {
+        if (_documentSynchronizer is not DefaultLSPDocumentSynchronizer documentSynchronizer)
+        {
+            Debug.Fail("Got an LSP document synchronizer I don't know how to handle.");
+            throw new InvalidOperationException("Got an LSP document synchronizer I don't know how to handle.");
```

Comment:
Is the Debug.Fail redundant if we're already throwing an exception here?

Reply:
`Debug.Fail` is intended to alert the developer with a message box, but has no effect at runtime. Having a different type of something in the MEF composition probably just means the developer making the new one made a typo or something, so the intent is to make it easier to find. This should really never happen.